### ISSUE TITLE
msys64 gcc compilation error

### ISSUE THREAD

**Issue by DamienLeBars** [capnproto/capnproto#1138]
06e471e introduced a compilation error with windows cygwin/msys gcc
```
capnproto/c++/src/kj/miniposix.c++:39:16: error: ‘_XOPEN_IOV_MAX’ was not declared in this scope
   39 |         return _XOPEN_IOV_MAX;
      |                ^~~~~~~~~~~~~~
```
(from [surelog compilation error in capnproto](https://github.com/alainmarcel/Surelog/runs/1583733628))

**Comment by DamienLeBars** [comment]
I found strange that the compilation configuration is not consistent for miniposix.c++ between:
capnproto/c++/Makefile.am (miniposix.c++ is missing)
and
capnproto/c++/src/kj/CMakeLists.txt

**Comment by kentonv** [comment]
`miniposix.c++` was added in #1129. Looks like @vlovich forgot to update `Makefile.am`, but this slipped past the tests because `miniposix.c++` isn't actually needed by any of the platforms we test on (it's a fallback for POSIX variants that don't define an appropriate constant).

But I am confused by your error report, because we actually do test on Cygwin and MinGW. In our test environment, either `IOV_MAX` or `UIO_MAX_IOV` (not sure which) is defined on Cygwin, and hence `miniposix.c++` is a no-op. On MinGW, `_WIN32` is defined, which also makes `miniposix.c++` a no-op. So how is it that this code is not being skipped in your case?

Are you actually targeting Cygwin, or MinGW? You say "cygwin/msys gcc", which is confusing -- my understanding is that MSYS is an alternative to Cygwin. MSYS is essentially a minimalist Unix shell environment intended to be used with MinGW GCC. So are you saying you're using MinGW GCC? But it's really weird that `_WIN32` isn't defined, then. If you're actually targeting Cygwin, then it makes sense that `_WIN32` is not defined, but one of the other constants should have been -- or is so in our test environment, at least.

I think possibly we should just delete the non-inline definition of `iovMax()` since we're not aware of any target systems that need it. I think I copy/pasted the code from somewhere long ago but I'm not sure if it has ever been exercised as part of this codebase. But we need to figure out why it's being compiled for you...

**Comment by DamienLeBars** [comment]
we target msys2 gcc (that is the same than cygwin), we use cmake and so capnproto/c++/src/kj/CMakeLists.txt is used (and so miniposix.c++ is compiled => with error)
in the cygwin compil test, miniposix.c++ is not compiled, but if you add it to capnproto/c++/Makefile.am, you will have an error.
I'm confused that there is a compilation system that is different for test.

**Comment by DamienLeBars** [comment]
ok, i think i understand one of my part of my problem.
For cygwin (or msys2) gcc compilation, i have to not use cmake that is only for visual studio, but the make only.
I havn't see that before (https://capnproto.org/install.html#prerequisites)

**Comment by kentonv** [comment]
Hi @DamienLeBars,

> we target msys2 gcc (that is the same than cygwin) and not mingw,

Sorry, I still don't understand what you mean here. MSYS2 is not a compiler. It is a shell environment built around the MinGW compiler. The MinGW compiler is not the same as the Cygwin compiler. MinGW is designed to compile programs built on the native Win32 APIs, whereas Cygwin provides a POSIX/Linux emulation layer. Programs built with the Cygwin compiler depend on the Cygwin runtime DLLs to operate; programs built with MinGW do not.

So which compiler are you using? Is it Cygwin, or is it MinGW?

> in the cygwin compil test, miniposix.c++ is not compiled, but if you add it to capnproto/c++/Makefile.am, you will have an error.

I don't see how that's possible. In `miniposix.h`, `iovMax()` is defined inline if either `IOV_MAX` or `UIO_MAX_IOV` is defined, and out-of-line otherwise. If it's declared out-of-line, and `miniposix.c++` is missing from the build, then we'd get a linker error. In our test Cygwin build, we are missing `miniposix.c++`, but we got no linger error, therefore we can conclude that either `IOV_MAX` or `UIO_MAX_IOV` must be defined. However, in this case, `miniposix.c++` should be a no-op, due to the `#ifdef`s.

**Comment by DamienLeBars** [comment]
MSYS2 is based on a modified version of Cygwin.
...
MSYS2 provides up-to-date native builds for GCC, mingw-w64, ....
[msys2](https://www.msys2.org/)
We use the GCC compiler that is a package of MSYS2.

**Comment by DamienLeBars** [comment]
I'm testing now only Cygwin auto compilation test here in my fork pf this repo. It seems that IOV_MAX is well defined when it is checked in the .h and not in the .c++ (i'm investigating)

**Comment by kentonv** [comment]
> I'm testing now only Cygwin auto compilation test here in my fork of this repo. It seems that IOV_MAX is well defined when it is checked in the .h and not in the .c++ (miniposix.c++ added in capnproto/c++/Makefile.am) (i'm investigating)

But `miniposix.c++` includes `miniposix.h`.

Maybe, though, all of the other files that include `miniposix.h` first include some other header that defines `IOV_MAX`? In that case we just need to add the missing include to `miniposix.h` I guess.

**Comment by kentonv** [comment]
It looks like the missing include might be `<sys/uio.h>`?

**Comment by kentonv** [comment]
No, that's not right, `<sys/uio.h>` is in there.

https://github.com/capnproto/capnproto/blob/06e471e75349174b10c735d5a2edbb1c8a158bc9/c++/src/kj/miniposix.h#L42-L44

**Comment by DamienLeBars** [comment]
yes, i'm trying to add necessary includes that is in the .h directly in the .c++
perhaps the ```#pragma once``` of the .h hide the define...
